// @mikro-orm/postgresql@7.1.5/PostgreSqlConnection.js
import { Pool, TypeOverrides } from 'pg';
import Cursor from 'pg-cursor';
import { PostgresDialect } from 'kysely';
import array from 'postgres-array';
import { AbstractSqlConnection, createPostgreSqlTypeParsers, Utils } from '@mikro-orm/sql';
/** PostgreSQL database connection using the `pg` driver. */
export class PostgreSqlConnection extends AbstractSqlConnection {
    createKyselyDialect(overrides) {
        const { onPoolCreated, ...poolOverrides } = (overrides ?? {});
        const options = this.mapOptions(poolOverrides);
        const pool = new Pool(options);
        void onPoolCreated?.(pool);
        return new PostgresDialect({
            pool,
            cursor: Cursor,
            onCreateConnection: this.options.onCreateConnection ?? this.config.get('onCreateConnection'),
            onReserveConnection: this.options.onReserveConnection ?? this.config.get('onReserveConnection'),
        });
    }
    mapOptions(overrides) {
        const ret = { ...this.getConnectionOptions() };
        const pool = this.config.get('pool');
        Utils.defaultValue(ret, 'max', pool?.max);
        Utils.defaultValue(ret, 'idleTimeoutMillis', pool?.idleTimeoutMillis);
        const types = new TypeOverrides();
        for (const [oid, parser] of Object.entries(createPostgreSqlTypeParsers(s => array.parse(s)))) {
            types.setTypeParser(Number(oid), parser);
        }
        ret.types = types;
        return Utils.mergeConfig(ret, overrides);
    }
    async callRoutine(routine, args = {}, ctx) {
        if (routine.type === 'function') {
            return this.callRoutineFunction(routine, args, ctx);
        }
        const quoted = (id) => this.platform.quoteIdentifier(id);
        const qualified = (routine.schema ? `${quoted(routine.schema)}.` : '') + quoted(routine.name);
        // Refcursor OUT params come back as server-generated cursor names to FETCH from later.
        const refcursorParams = routine.params.filter(p => p.direction !== 'in' && typeof p.type === 'string' && /^refcursor$/i.test(p.type));
        // Refcursors are transaction-scoped — fail fast instead of opening cursors we can't read.
        if (refcursorParams.length > 0 && !ctx) {
            throw new Error(`Routine ${routine.name} declares refcursor OUT params on PostgreSQL but was not called inside a transaction. Wrap the call in 'em.transactional(...)' so the refcursor OUT params remain valid for FETCH.`);
        }
        const placeholders = routine.params.map(() => '?').join(', ');
        const positional = routine.params.map(p => this.convertRoutineInbound(args[p.name], p));
        const rows = (await this.execute(`call ${qualified}(${placeholders})`, positional, 'all', ctx));
        const row = rows[0] ?? {};
        const scalarOutParams = routine.params.filter(p => p.direction !== 'in' && !refcursorParams.includes(p));
        this.applyRoutineOutParams(row, scalarOutParams, args);
        if (refcursorParams.length > 0) {
            return (await this.fetchRefcursors(row, routine, refcursorParams, ctx));
        }
        return undefined;
    }
    async fetchRefcursors(row, routine, refcursorParams, ctx) {
        const cursorNames = refcursorParams
            .map(p => row[p.name])
            .filter((name) => typeof name === 'string');
        const sets = [];
        for (const cursorName of cursorNames) {
            const fetched = (await this.execute(`fetch all from "${cursorName.replaceAll('"', '""')}"`, [], 'all', ctx));
            sets.push(fetched);
        }
        return sets;
    }
}
