// @mikro-orm/sql@7.1.5/SqlEntityManager.js
import { EntityManager, raw, Utils, } from '@mikro-orm/core';
import { MikroKyselyPlugin } from './plugin/index.js';
/**
 * @inheritDoc
 */
export class SqlEntityManager extends EntityManager {
    /**
     * Creates a QueryBuilder instance
     */
    createQueryBuilder(entityName, alias, type, loggerContext) {
        const context = this.getContext(false);
        const qb = this.driver.createQueryBuilder(entityName, context.getTransactionContext(), type, true, loggerContext ?? context.loggerContext, alias, this);
        qb.setAbortOptions(context.getAbortOptions());
        return qb;
    }
    /**
     * Shortcut for `createQueryBuilder()`
     */
    qb(entityName, alias, type, loggerContext) {
        return this.createQueryBuilder(entityName, alias, type, loggerContext);
    }
    /**
     * Returns a configured Kysely instance bound to this EntityManager.
     *
     * When the EntityManager is inside a transaction (e.g. within `em.transactional(...)`, or after
     * `em.begin()`), the returned Kysely instance automatically uses the transaction context, so any
     * queries executed via Kysely's own `.execute()` / `.executeTakeFirst*()` participate in the
     * current transaction.
     *
     * If you need a Kysely instance that is **not** bound to the current transaction (e.g. to perform
     * a side query against the pool while inside a transactional block), fork the EntityManager first:
     *
     * ```ts
     * await em.transactional(async em => {
     *   // bound to the current transaction
     *   await em.getKysely().selectFrom('user').selectAll().execute();
     *
     *   // bound to the pool, runs outside the transaction
     *   await em.fork().getKysely().selectFrom('audit_log').selectAll().execute();
     * });
     * ```
     *
     * The `options.type` (`'read'` / `'write'`) is only honored outside a transaction — inside a
     * transaction the connection is already pinned.
     */
    getKysely(options = {}) {
        const context = this.getContext(false);
        const ctx = context.getTransactionContext();
        let kysely = ctx ?? this.getConnection(options.type).getClient();
        if (options.columnNamingStrategy != null ||
            options.tableNamingStrategy != null ||
            options.processOnCreateHooks != null ||
            options.processOnUpdateHooks != null ||
            options.convertValues != null) {
            kysely = kysely.withPlugin(new MikroKyselyPlugin(this, options));
        }
        return kysely;
    }
    async execute(query, params = [], methodOrOptions = 'all', loggerContext) {
        const opts = typeof methodOrOptions === 'string' ? { method: methodOrOptions, loggerContext } : methodOrOptions;
        const context = this.getContext(false);
        // Per-field fallback to fork-level abort, matching `EntityManager.prepareOptions` semantics.
        const fork = context.getAbortOptions();
        const hasAbort = opts.signal != null || opts.inflightQueryAbortStrategy != null || fork != null;
        // Connection layer carries abort piggy-backed on `loggerContext`; merge only when needed.
        const merged = hasAbort || opts.loggerContext ? { ...opts.loggerContext } : undefined;
        if (merged && hasAbort) {
            merged.signal = opts.signal ?? fork?.signal;
            merged.inflightQueryAbortStrategy = opts.inflightQueryAbortStrategy ?? fork?.inflightQueryAbortStrategy;
        }
        return this.getDriver().execute(query, params, opts.method ?? 'all', context.getTransactionContext(), merged);
    }
    /**
     * @inheritDoc
     */
    async countBy(entityName, groupBy, options = {}) {
        const em = this.getContext(false);
        options = { ...options };
        em.prepareOptions(options);
        const meta = em.getMetadata().find(entityName);
        const fields = Utils.asArray(groupBy);
        const { where: rawWhere, ...countOptions } = options;
        await em.tryFlush(entityName, options);
        const where = await em.processWhere(entityName, rawWhere ?? {}, options, 'read');
        const qb = em.createQueryBuilder(meta.class);
        qb
            .select([...fields, raw('count(*) as cnt')])
            .where(where)
            .groupBy(fields);
        if (countOptions.having) {
            qb.having(countOptions.having);
        }
        if (countOptions.schema) {
            qb.withSchema(countOptions.schema);
        }
        const rows = await qb.execute('all', { mapResults: false });
        const results = {};
        for (const row of rows) {
            const keyParts = fields.map(f => {
                const col = meta.properties[f]?.fieldNames?.[0] ?? f;
                return String(row[col]);
            });
            const key = keyParts.join(Utils.PK_SEPARATOR);
            results[key] = +row.cnt;
        }
        return results;
    }
    getRepository(entityName) {
        return super.getRepository(entityName);
    }
    applyDiscriminatorCondition(entityName, where) {
        // this is handled in QueryBuilder now for SQL drivers
        return where;
    }
}
