// @mikro-orm/sql@7.1.5/query/enums.js
/** Type of SQL query to be generated. */
export var QueryType;
(function (QueryType) {
    QueryType["TRUNCATE"] = "TRUNCATE";
    QueryType["SELECT"] = "SELECT";
    QueryType["COUNT"] = "COUNT";
    QueryType["INSERT"] = "INSERT";
    QueryType["UPDATE"] = "UPDATE";
    QueryType["DELETE"] = "DELETE";
    QueryType["UPSERT"] = "UPSERT";
})(QueryType || (QueryType = {}));
/** Operators that apply to the embedded array column itself, not to individual elements. */
export const EMBEDDABLE_ARRAY_OPS = ['$contains', '$contained', '$overlap'];
/** Type of SQL JOIN clause. */
export var JoinType;
(function (JoinType) {
    JoinType["leftJoin"] = "left join";
    JoinType["innerJoin"] = "inner join";
    JoinType["nestedLeftJoin"] = "nested left join";
    JoinType["nestedInnerJoin"] = "nested inner join";
    JoinType["pivotJoin"] = "pivot join";
    JoinType["innerJoinLateral"] = "inner join lateral";
    JoinType["leftJoinLateral"] = "left join lateral";
})(JoinType || (JoinType = {}));
